File: backend/test/model/model.test.ts
import { describe, it, expect } from 'vitest';
import { Decimal } from 'decimal.js';
import { Asset } from '../../src/model/asset';

describe('ModelTest', () => {
  it('testAssetMappings', () => {
    const btc = new Asset(Asset.BTC);
    expect(btc.krakenTicker).toBe(Asset.XBT);
    expect(btc.tradingPair).toBe('XBTUSD');
    expect(btc.isUsd).toBe(false);

    const doge = new Asset(Asset.DOGE);
    expect(doge.krakenTicker).toBe(Asset.XDG);
    expect(doge.tradingPair).toBe('XDGUSD');

    const usd = new Asset(Asset.USD);
    expect(usd.isUsd).toBe(true);
    expect(usd.tradingPair).toBe('USDUSD');

    const eth = new Asset(Asset.ETH);
    expect(eth.krakenTicker).toBe(Asset.ETH);
    expect(eth.tradingPair).toBe('ETHUSD');

    expect(Asset.toKrakenTicker('btc')).toBe(Asset.XBT);
    expect(Asset.toKrakenTicker('doge')).toBe(Asset.XDG);
    expect(Asset.toKrakenTicker('eth')).toBe('ETH');

    expect(Asset.tradingPair('btc')).toBe('XBTUSD');
    expect(Asset.tradingPair('eth')).toBe('ETHUSD');

    expect(Asset.BTC_USD_PAIR).toBe('XBTUSD');
  });

  it('testPortfolioSnapshot', () => {
    const asset = {
      symbol: Asset.BTC,
      balance: new Decimal(1.0),
      price: new Decimal(10.0),
      valueUSD: new Decimal(10.0),
      targetPercent: new Decimal(1.0),
      currentPercent: new Decimal(1.0),
      deviationPercent: new Decimal(0.0),
      deviationUSD: new Decimal(0.0)
    };

    expect(asset.symbol).toBe(Asset.BTC);

    const snapshot = {
      timestamp: new Date(0).toISOString(),
      totalValueUSD: new Decimal(10.0),
      assets: { [Asset.BTC]: asset },
      actions: ['BUY'],
      drawdownPercent: new Decimal(0.0),
      fiatDeploymentPercent: new Decimal(0.0),
      effectiveUsdTargetPercent: new Decimal(0.0)
    };

    expect(snapshot.totalValueUSD.toNumber()).toBe(10.0);
  });
});

File: backend/src/test/kotlin/com/gemini/krakenbot/model/ModelTest.kt
package com.gemini.krakenbot.model

import com.gemini.krakenbot.TestFixtures
import com.gemini.krakenbot.domain.CASH_RESERVE_FACTOR
import com.gemini.krakenbot.domain.HUNDRED
import com.gemini.krakenbot.util.PrecisionConstants
import io.kotest.core.spec.IsolationMode
import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.comparables.shouldBeEqualComparingTo
import io.kotest.matchers.shouldBe
import java.math.BigDecimal

class ModelTest : StringSpec() {
    override fun isolationMode() = IsolationMode.InstancePerTest

    init {
        "testAssetMappings" {
            val btc = Asset(Asset.BTC)
            btc.krakenTicker shouldBe Asset.XBT
            btc.tradingPair shouldBe TestFixtures.XBTUSD
            btc.isUsd shouldBe false

            val doge = Asset(Asset.DOGE)
            doge.krakenTicker shouldBe Asset.XDG
            doge.tradingPair shouldBe "XDGUSD"

            val usd = Asset(Asset.USD)
            usd.isUsd shouldBe true
            usd.tradingPair shouldBe "USDUSD"

            val eth = Asset(Asset.ETH)
            eth.krakenTicker shouldBe Asset.ETH
            eth.tradingPair shouldBe TestFixtures.ETHUSD

            Asset.toKrakenTicker(TestFixtures.BTC_LOWER) shouldBe Asset.XBT
            Asset.toKrakenTicker("doge") shouldBe Asset.XDG
            Asset.toKrakenTicker(TestFixtures.ETH_LOWER) shouldBe "ETH"

            Asset.tradingPair(TestFixtures.BTC_LOWER) shouldBe TestFixtures.XBTUSD
            Asset.tradingPair(TestFixtures.ETH_LOWER) shouldBe TestFixtures.ETHUSD

            Asset.BTC_USD_PAIR shouldBe TestFixtures.XBTUSD
        }

        "matchesUsdQuotedPair accepts exact aliases and rejects substring collisions" {
            Asset.matchesUsdQuotedPair("ETHUSD", Asset.ETH) shouldBe true
            Asset.matchesUsdQuotedPair("XETHZUSD", Asset.ETH) shouldBe true
            Asset.matchesUsdQuotedPair("SOMETHINGETHUSD", Asset.ETH) shouldBe false
            Asset.matchesUsdQuotedPair("XBTUSD", Asset.BTC) shouldBe true
            Asset.matchesUsdQuotedPair("XXBTZUSD", Asset.BTC) shouldBe true
            Asset.matchesUsdQuotedPair("XBTUSDT", Asset.BTC) shouldBe false
        }

        "testServiceCompanions" {
            PrecisionConstants.CASH_RESERVE_FACTOR.shouldBeEqualComparingTo(BigDecimal("0.99"))

            PrecisionConstants.HUNDRED.shouldBeEqualComparingTo(BigDecimal("100"))
            PrecisionConstants.SCALE_PERCENT shouldBe 4
            PrecisionConstants.SCALE_CRYPTO shouldBe 8
            PrecisionConstants.SCALE_USD shouldBe 2
        }
    }
}
